### PR TITLE
fix: non-deterministic CI test failure

### LINKED ISSUE #389: `One of the worker has failed` error in test
## 🐛 Bug

When CI tests are run, a non-deterministic `RuntimeError: One of the worker has failed` keep coming.

For reference:
- https://github.com/Lightning-AI/litdata/actions/runs/11139472937/job/30956253045
- https://github.com/Lightning-AI/litdata/actions/runs/10898567830/job/30249673425?pr=348
- https://github.com/Lightning-AI/litdata/actions/runs/10905566383/job/30264717145

This is a standard `race condition` issue. 

If the optimize function is too small, then instead of waiting & checking for error_queue, we simply assume that a worker has failed.

https://github.com/Lightning-AI/litdata/blob/main/src/litdata/processing/data_processor.py#L1140

---

### Possible fix 🛠️

IMO, to fix this, we should wait for 0.01 seconds; if no error occurs, everything is fine.

```python
 if all(not w.is_alive() for w in self.workers):
    try:
        error = self.error_queue.get(timeout=0.001)
         self._exit_on_error(error)
    except Empty:
         break
```

---

Thoughts?

cc: @tchaton

### PR DESCRIPTION
**Pull request by deependujha**
<details>
  <summary><b>Before submitting</b></summary>

- [ ] Was this discussed/agreed via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lit-data/blob/main/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

</details>

## What does this PR do?

Fixes #389 

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in GitHub issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -28,3 +28,10 @@ clean:
 	rm -rf ./src/*.egg-info
 	rm -rf ./build
 	rm -rf ./dist
+
```
Comment (deependujha) [inline Makefile:31]: for quick installation and setup. Might be helpful in `contributing.md` PR

File: src/litdata/processing/data_processor.py [hunk 2]
```diff
@@ -1135,10 +1135,15 @@ def run(self, data_recipe: DataRecipe) -> None:
...
+            # This means either there were some kinda of errors, or optimize function was very small.
             if all(not w.is_alive() for w in self.workers):
-                raise RuntimeError("One of the worker has failed")
+                for w in self.workers:
+                    if w.exitcode != 0:
+                        try:
+                            error = self.error_queue.get(timeout=0.01)
+                            self._exit_on_error(error)
+                        except Empty:
+                            continue
```
Comment (deependujha) [inline src/litdata/processing/data_processor.py:1146]: is this what you meant @tchaton ?

I think we can remove the for loop and simply wait for error_queue if any error is there!

File: Makefile [hunk 3]
```diff
@@ -28,3 +28,10 @@ clean:
...
+install:
+	pip install -r requirements.txt
+	pip install -r requirements/test.txt
+	pip install -r requirements/docs.txt
+	pip install -r requirements/extras.txt
+	pip install -e .
```
Comment (deependujha) [inline Makefile:37]: for quick installation of dependencies, and might be helpful in `contributing.md` PR.

File: src/litdata/processing/data_processor.py (hunk 2 above)
Comment (tchaton) [reply src/litdata/processing/data_processor.py:1146]: Yes, this should work too.

**Review by Borda — APPROVED** [review]

**Pull request merged**